### ISSUE TITLE
[QST] libcudf `fixed_point` Precision Support?

### ISSUE THREAD

**Issue by codereport** [rapidsai/cudf#6656]
@kkraus14, @shwina and I met today to discuss cudf Python support for DecimalType (aka bindings to the libcudf `fixed_point`). One of the questions that came up was whether the libcudf `fixed_point` should support the `precision` parameter that is supported by DecimalType APIs more common outside of the C++ ecosystem. For example:

* [pyarrow.Decimal128Type](http://arrow.apache.org/docs/python/generated/pyarrow.Decimal128Type.html#pyarrow.Decimal128Type)
* [Spark DecimalType](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/types/DecimalType.html#DecimalType-int-int-)
* [Transact-SQL](https://docs.microsoft.com/en-us/sql/t-sql/data-types/decimal-and-numeric-transact-sql?view=sql-server-ver15)

While there seems to be [good documentation](https://docs.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-ver15#:~:text=Precision%20is%20the%20number%20of,decimal%20data%20types%20is%2038) for how to calculate scale/precision for the SQL type, there isn't for the pyarrow (which seems to have issues with it's implementation). I will look into all of the above to get a better understanding of their behaviour.

**The question is what to do on the libcudf side to provide support?** The two main alternatives (coming out of the conversation with Keith and Ashwin) seemed to be:

1. Add precision support to the libcudf `fixed_point`
2. Add libcudf functions for calculate what precision _would be_ and expose this so Python (and other bindings) can use it to handle precision and type demotion/promotion on it's side

I lean towards option 2 as it keeps the C++ `fixed_point` more consistent with CNL and other `fixed_point` implementations in the C++ ecosystem. However, the point of this issue is to open a discussion.

**Comment by kkraus14** [comment]
cc @jlowe @revans2 for Spark related discussion
cc @felipeblazing @williamBlazing for BlazingSQL

From my perspective, the main reason to support a `precision` parameter is to allow having a higher level logical `decimal` type that we can control the physical storage underneath based on the computed `precision` / `scale`. This is how Transact-SQL does things above, but seems to differ from Arrow and Spark which seem to use constant 128bit storage per element.

I agree with @codereport in that I lean towards option 2.

**Comment by jrhemstad** [comment]
Can someone remind me what `precision` means?

**Comment by kkraus14** [comment]
`precision` is defined as the total number of digits allowed in the `decimal` type. Both the digits to the left and right of the decimal point.

**Comment by jrhemstad** [comment]
So `precision` just boils down to choosing between `decimal32` and `decimal64` and ensuring you have enough bits to support the number of digits. 

I don't know what it would mean to "add precision support to libcudf" (option 1). 

If a precision of `N` is required, you use the smallest of `decimal32` or `decimal64` that satisfies `N` (or throw an error if `N` is larger than what can be supported by `decimal64`).  Is option 2 just providing a way to query what the maximum `N` is supported by `decimal32` and `decimal64`?

**Comment by kkraus14** [comment]
The question is mostly in the context of unary / binary ops which change the `precision` / `scale` based on the operation. I.E. multiply two numbers increases both the precision and scale.

From the libcudf side we could choose to similarly have a logical `decimal` type which automatically handles the underlying physical type promotion as needed to store the logical `decimal` type. For example if I multiplied two `decimal` types that fit in a 32bit space such that the resulting `decimal` type no longer fits in a 32bit space (calculated based on `precision` and `scale`), we could automatically handle that in libcudf versus the workaround in Python would likely be to typecast the two input columns to a  64bit space before doing the operation.

**Comment by jrhemstad** [comment]
> From the libcudf side we could choose to similarly have a logical `decimal` type which automatically handles the underlying physical type promotion as needed to store the logical `decimal` type.

Not a fan of this idea as it would be unlike any other type we have and would cause a great deal of complexity and headache. In my mind, it's no different than the fact that we have separate `INT8, INT16, INT32, INT64` types. We could instead have a single logical `INT` type and play games with the underlying storage, but that would be a huge headache so we don't do it.

> For example if I multiplied two decimal types that fit in a 32bit space such that the resulting decimal type no longer fits in a 32bit space (calculated based on precision and scale), we could automatically handle that in libcudf versus the workaround in Python would likely be to typecast the two input columns to a 64bit space before doing the operation.

This again seems the same as how we handle normal integers. If you multiply two `INT32` values, we don't do any checking to ensure the result doesn't overflow into an `INT32` and automagically upconvert to `INT64`.

**Comment by jlowe** [comment]
The Spark plugin should be able to handle managing the precision tracking without needing libcudf to do this automatically, similar to Jake's analogy of the INT32 multiply example.

> versus the workaround in Python would likely be to typecast the two input columns to a 64bit space before doing the operation

Can we specify `DECIMAL64` and the desired scale in the binaryop output type and expect libcudf to do the right thing during the operation?  Would be nice to avoid manifesting two cast columns.

**Comment by kkraus14** [comment]
> Can we specify `DECIMAL64` and the desired scale in the binaryop output type and expect libcudf to do the right thing during the operation?

I thought there's specific rules for calculating output `precision` and `scale` that should be followed.

> This again seems the same as how we handle normal integers. If you multiply two INT32 values, we don't do any checking to ensure the result doesn't overflow into an INT32 and automagically upconvert to INT64.

The difference here is we don't need to introspect the data. I.E. if I have two `DECIMAL32` types with `precision` of 3, then I know that the result `precision` is 7 which can still fit in a `DECIMAL32` type as opposed to needed a `DECIMAL64` type.

**Comment by harrism** [comment]
I am on the fence on this one, FWIW. I don't see it as exactly analogous to `INT` types, because in this case the ecosystem is used to large (`DECIMAL128`) decimal types that automatically handle the precision. If every user of libcudf decimals is going to have to implement the result [precision/scale table from this informative page](https://docs.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-ver15#:~:text=Precision%20is%20the%20number%20of,decimal%20data%20types%20is%2038), and then decide when to cast their inputs in order to accomodate the increased precision, then why shouldn't libcudf do that calculation for them?.

**Comment by jrhemstad** [comment]
> the ecosystem is used to large (`DECIMAL128`) decimal types that automatically handle the precision.

The ecosystem (Pandas) is also used to things like when you do a sum reduction on an INT8 column it automatically up-converts to INT64 to avoid overflow, or being able to use floating point values as a gather map. libcudf Is Not Pandas (LINP) and doesn't do any of those things because it's not consistent with C++. It's been the design philosophy of libcudf to not try and do _everything_ and instead provide the necessary primitives to enable consumers (cuDF/Spark/Blazing) to do whatever they need for their users. 

> decide when to cast their inputs in order to accommodate the increased precision, then why shouldn't libcudf do that calculation for them?

Because that's a significant deviation from the current binary ops behavior. The current API is a garbage in, garbage out design, i.e., it uses whatever the lhs/rhs and specified output types are and doesn't make any judgement about what those types are (other than disallowing obvious things like `string` + `int`). 

I do not think we should start trying to do automagic type/precision handling for decimal types _in libcudf_. I fully support identifying the necessary primitives to enable libcudf consumers to avoid needing to reinvent the wheel as much as possible. I don't know what those primitives are yet, perhaps codifying the information from this [page](https://docs.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-ver15#:~:text=Precision%20is%20the%20number%20of,decimal%20data%20types%20is%2038) in libcudf so that it can be queried programmatically.

**Comment by kkraus14** [comment]
> I do not think we should start trying to do automagic type/precision handling for decimal types _in libcudf_. I fully support identifying the necessary primitives to enable libcudf consumers to avoid needing to reinvent the wheel as much as possible. I don't know what those primitives are yet, perhaps codifying the information from this [page](https://docs.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-ver15#:~:text=Precision%20is%20the%20number%20of,decimal%20data%20types%20is%2038) in libcudf so that it can be queried programmatically.

Just one point: libcudf is already doing automagic `scale` handling for decimal types _in libcudf_ because there's defined rules to do so. Presumably we could do the same thing for `precision` if desired.

We can punt on the discussion of changing the underlying storage type (decimal32 vs decimal64 vs decimal128, etc.) to the future. If libcudf wanted to support the functions for calculating the precision of different operations that would be great and we can prototype the type promotion / demotion from the cuDF Python side to guide future discussion.

**Comment by codereport** [comment]
TLDR: We decided to code precision logic on Python/Spark side of things and not have a notion of precision in libcudf's `fixed_point`.

You can see the work done in https://github.com/rapidsai/cudf/pull/6715, specifically:https://github.com/rapidsai/cudf/blob/26c2dfed5654c5a093df2cba7047f1e43eabfa71/python/cudf/cudf/core/column/decimal.py#L73
